lib/ui/painting/canvas.cc

```diff
@@ -639,6 +640,77 @@ void Canvas::drawShadow(const CanvasPath* path,
   }
 }
 
+Dart_Handle Canvas::drawGlyphRun(Dart_Handle glyphs_handle,
+                                 Dart_Handle positions_handle,
+                                 double originX,
+                                 double originY,
+                                 Dart_Handle font_family_name_handle,
+                                 int font_weight_index,
+                                 int font_slant_index,
+                                 double font_size,
+                                 Dart_Handle paint_objects,
+                                 Dart_Handle paint_data) {
+  Paint paint(paint_objects, paint_data);
+  FML_DCHECK(paint.isNotNull());
+  if (display_list_builder_) {
+    tonic::Uint16List glyphs(glyphs_handle);
+    tonic::Float32List positions(positions_handle);
+    std::string font_family_name =
+        tonic::StdStringFromDart(font_family_name_handle);
+    auto sk_glyphs = reinterpret_cast<const SkGlyphID*>(glyphs.data());
+    auto sk_points = reinterpret_cast<const SkPoint*>(positions.data());
+
+    // Convert txt::FontWeight values (ranging from 0-8) to SkFontStyle:Weight
+    // values (ranging from 100-900)
+    SkFontStyle::Weight font_weight =
+        static_cast<SkFontStyle::Weight>(font_weight_index * 100 + 100);
+    // Convert txt::FontStyle to SkFontFontStyle::Slant
+    SkFontStyle::Slant font_slant = font_slant_index == 0
+                                        ? SkFontStyle::Slant::kUpright_Slant
+                                        : SkFontStyle::Slant::kItalic_Slant;
+    SkFontStyle font_style(font_weight, SkFontStyle::Width::kNormal_Width,
+                           font_slant);
+
+    auto font_collection = UIDartState::Current()
+                               ->platform_configuration()
+                               ->client()
+                               ->GetFontCollection()
+                               .GetFontCollection();
+
+    std::vector<SkString> sk_font_families;
+    sk_font_families.push_back(SkString(font_family_name.c_str()));
+
+    auto sk_typefaces =
+        font_collection->CreateSktFontCollection()->findTypefaces(
+            sk_font_families, font_style);
+
+    if (sk_typefaces.empty()) {
```

Comment:
@jason-simmons or @Rusino - will this be the case if the exact family/size/weight/etc. cannot be found, or will Skia potentially fall back here to some other font?

There also probably needs to be some kind of assertion that you've found only one font right? This is important so that the glyph IDs will actually match what you expected.

Reply:
SkParagraph's `FontCollection::findTypefaces` may return some other font.  For example, if none of the available font managers can find a match for the specified font family, then `findTypefaces` may return the platform's default font.

I'd be reluctant to expose a public API that uses glyph IDs without a more complete set of low-level font management APIs similar to Skia's `SkFontMgr`/`SkTypeface`/`SkFont`.  These APIs would need to offer more control over font selection and a way to map Unicode code points to glyph IDs.

Reply:
I agree. If you're drawing with glyph IDs it will be very important to know you're getting the font you expect. 

For others reading this who aren't familiar, we get a number of bugs from users related to unexpected fallbacks/mismatched versions of fonts as it is. Sometimes a particular font has a bug and does not render as expected. It tends to be difficult today to track these down and understand them. That problem would be significantly worse with an API like this. If you're supplying glyph IDs based on the font on your local development environment, and your user has a different font installed, you suddenly will render either nothing or possibly _completely different text_ (because locally the glyph ID for `A` is now `B` on the installed font).